### PR TITLE
python-language-server: 2020-06-19 -> 2020-10-08

### PR DESCRIPTION
**Pull request by futile**
###### Motivation for this change

This updates to a version of `python-language-server` that supports Python 3.9 (which is also the current `master` of `github:microsoft/python-language-server` at the time of writing).

@thomasjm Pinging you since you are the maintainer; I ran `create-deps.sh` and then added all package versions that were missing during `nix-build` to `manual_deps.txt` until the build succeeded. Was that right?

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Before:
```
/nix/store/s3vmrmpmliwy0sxfpzr7cp0ygkk6xxzc-python-language-server-2020-06-19	  165113576
```
After:
```
/nix/store/2k67mrg6kc690h1m898ga3dryslrcgl3-python-language-server-2020-10-08	  165204600
```

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 121522` at f68ecfc7 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>python-language-server</li>
  </ul>
</details>
<details>
  <summary>2 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`buildPhase` should probably contain `runHook preBuild` and `runHook postBuild`.

Near pkgs/development/dotnet-modules/python-language-server/default.nix:64:3:
```
   |
64 |   buildPhase = ''
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/development/dotnet-modules/python-language-server/default.nix:80:3:
```
   |
80 |   installPhase = ''
   |   ^
```
</li>
  </ul>
</details>

**Comment by thomasjm** [comment]
> added all package versions that were missing during nix-build to manual_deps.txt until the build succeeded. Was that right?

Yep, that sounds good. It would be good to also remove older versions of things in `manual_deps.txt` that aren't needed to build anymore, but not crucial.

If you tested it and it works then LGTM, thanks!

**Comment by futile** [comment]
> > added all package versions that were missing during nix-build to manual_deps.txt until the build succeeded. Was that right?
> 
> Yep, that sounds good. It would be good to also remove older versions of things in `manual_deps.txt` that aren't needed to build anymore, but not crucial.
> 
> If you tested it and it works then LGTM, thanks!

Thanks! Do you know a good way to do so? Then I could just do that.

What do you think about including the `pre`/`post`-hooks mentioned by the bot? If those make sense for the package I could just add them as well.

**Comment by thomasjm** [comment]
Unfortunately the only way I know to do so is to try removing one and see if it still builds...

Adding the hooks sounds fine to me if it still builds and works after, I think the only reason they aren't there is because that bot didn't exist when I added this originally.

**Comment by futile** [comment]
Well, turns out it still builds without all the old deps! I also tested the binary and it still works, so I'm not sure if removing all the other deps could cause something to go wrong at runtime, but it builds just fine.

I also added the `runHook`s recommended by the bot, and made the script a lil bit safer by quoting some paths and adding `set -euo pipefail` so the script aborts on errors, unknown variables and on errors in piped statements ([link](https://coderwall.com/p/fkfaqq/safer-bash-scripts-with-set-euxo-pipefail)).

**Review by thomasjm — APPROVED** [review]
Huh, it's slightly concerning that so many manual deps aren't needed anymore, but I built it myself and it seems to work.

All the other changes look great, thanks!

**Review by thomasjm — CHANGES_REQUESTED** [review]
Err actually, I just read over the discussion on the initial PR and I think there is the possibility of runtime issues due to those missing deps. I think we should hold off until it's been tested a little more thoroughly at runtime. I haven't had a chance to try yet.

PR here: https://github.com/NixOS/nixpkgs/pull/85930

**Comment by futile** [comment]
> Huh, it's slightly concerning that so many manual deps aren't needed anymore, but I built it myself and it seems to work.
> 
Yeah, I thought the same, the build passing with almost all of the manual deps removed is kind of suspicious. I didn't build on darwin/macos, maybe some of them are needed there?

> All the other changes look great, thanks!

Thanks! :)





> Err actually, I just read over the discussion on the initial PR and I think there is the possibility of runtime issues due to those missing deps. I think we should hold off until it's been tested a little more thoroughly at runtime. I haven't had a chance to try yet.
> 
> PR here: #85930

Ah, that seems helpful! Though the old discussion is kind of hard to follow for me. What would you suggest as a way forward? I'll keep running this PR as my `python-language-server` since I need Python 3.9 support, so I might be able to spot/debug issues that pop up. I could also just add in the dependencies I took out, but maybe there are new ones missing after the update. I'm not really familiar with dotnet development, so don't really know how (build-&runtime-)dependencies work.

**Comment by thomasjm** [comment]
> What would you suggest as a way forward? I'll keep running this PR as my python-language-server since I need Python 3.9 support, so I might be able to spot/debug issues that pop up.

I'm not a dotnet expert either unfortunately, the deps were added in a sort of scattershot way. There was at least one case where I noticed something didn't work at runtime until one was added -- it was a dependency on libicu via openssl.

But, I'd say that if you're actually using it and the basic stuff is working (completion, symbols, hovers, etc.) then it should be good to go (you're on Linux I assume?). It would be good to do a smoketest on MacOS before merging to be safe though.

**Comment by thomasjm** [comment]
I don't recall how I triggered the `libicu` thing, but it was during basic usage. If the basic functionality is working for you then I think we're good.

Hopefully someone responds about darwin, if not I can try to get access to a mac.

**Comment by alyssais** [comment]
@ofborg build python-language-server

**Comment by futile** [comment]
As @alyssais just pointed out to me, this is actually not supported on darwin anyway, so don't need to get a darwin test. (Thanks @alyssais for taking a look!)

I guess this means the only thing missing is testing a bit more during regular use?

**Review by thomasjm — APPROVED** [review]

**Comment by thomasjm** [comment]
Right good point--this looks good then, a little more usage testing would be welcome but merging now sounds fine to me. Thanks all for checking.

**Review by jonringer — APPROVED** [review]
LGTM

doesn't crash when starting it...

```
https://github.com/NixOS/nixpkgs/pull/121522

1 package built:
python-language-server
```

**Comment by futile** [comment]
@jonringer thanks for review! Anything else necessary before this can be merged?

**Comment by jonringer** [comment]
No, I just forgot to click merge :)
